File: backend/src/lista.cpp
#include "../include/lista.hpp"

// 1. Inicializamos la lista vacía
Lista::Lista() {
    cabeza = nullptr;
    cantidad = 0;
}

// 2. Insertar un elemento al final de la lista
void Lista::insertar(int dato) {
    NodoLista* nuevoNodo = new NodoLista(dato);

    if (estaVacia()) {
        // Si la lista estaba vacía, este nodo pasa a ser la cabeza
        cabeza = nuevoNodo;
    } else {
        // Recorremos hasta el último nodo para engancharlo ahí
        NodoLista* actual = cabeza;
        while (actual->siguiente != nullptr) {
            actual = actual->siguiente;
        }
        actual->siguiente = nuevoNodo;
    }
    cantidad++;
}

// 3. Eliminar la primera ocurrencia de un valor
bool Lista::eliminar(int dato) {
    if (estaVacia()) {
        return false;
    }

    // Caso especial: el dato está en la cabeza
    if (cabeza->dato == dato) {
        NodoLista* aux = cabeza;
        cabeza = cabeza->siguiente;
        delete aux;
        cantidad--;
        return true;
    }

    // Buscamos el nodo anterior al que queremos eliminar
    NodoLista* anterior = cabeza;
    while (anterior->siguiente != nullptr && anterior->siguiente->dato != dato) {
        anterior = anterior->siguiente;
    }

    // Si llegamos al final sin encontrarlo, no existe en la lista
    if (anterior->siguiente == nullptr) {
        return false;
    }

    // "Saltamos" el nodo a eliminar para desenlazarlo
    NodoLista* aux = anterior->siguiente;
    anterior->siguiente = aux->siguiente;
    delete aux;
    cantidad--;
    return true;
}

// 4. Eliminar un elemento según su posición (0 = cabeza)
bool Lista::eliminarPorPosicion(int pos) {
    if (estaVacia() || pos < 0 || pos >= cantidad) {
        return false;
    }

    if (pos == 0) {
        NodoLista* aux = cabeza;
        cabeza = cabeza->siguiente;
        delete aux;
        cantidad--;
        return true;
    }

    NodoLista* anterior = cabeza;
    for (int i = 0; i < pos - 1; i++) {
        anterior = anterior->siguiente;
    }

    NodoLista* aux = anterior->siguiente;
    anterior->siguiente = aux->siguiente;
    delete aux;
    cantidad--;
    return true;
}

// 5. Verificar si está vacía
bool Lista::estaVacia() {
    return cabeza == nullptr;
}

// 6. Cantidad de elementos actuales
int Lista::tamano() {
    return cantidad;
}

// 7. Verificar si un valor existe en la lista
bool Lista::contiene(int dato) {
    NodoLista* actual = cabeza;
    while (actual != nullptr) {
        if (actual->dato == dato) {
            return true;
        }
        actual = actual->siguiente;
    }
    return false;
}

// 8. Entregar todos los valores en orden para que la web los dibuje
//    como una cadena de cajas conectadas por flechas.
std::vector<int> Lista::obtenerElementos() {
    std::vector<int> elementos;
    NodoLista* actual = cabeza;
    while (actual != nullptr) {
        elementos.push_back(actual->dato);
        actual = actual->siguiente;
    }
    return elementos;
}

// 9. Destructor: limpia toda la memoria si se destruye la lista
Lista::~Lista() {
    while (!estaVacia()) {
        eliminarPorPosicion(0);
    }
}

File: frontend/js/lista.js
/**
 * Lista Enlazada Simple - Portado a JavaScript desde backend/src/lista.cpp
 * (Persona 3) para poder ejecutarse en el navegador dentro del simulador web.
 */
class Lista {
    constructor() {
        this.cabeza = null; // NodoLista* cabeza
        this.cantidad = 0;
    }

    // Insertar al final
    insertar(dato) {
        const estabaVacia = this.estaVacia();
        const nuevoNodo = { dato, siguiente: null };

        if (estabaVacia) {
            this.cabeza = nuevoNodo;
        } else {
            let actual = this.cabeza;
            while (actual.siguiente !== null) {
                actual = actual.siguiente;
            }
            actual.siguiente = nuevoNodo;
        }
        this.cantidad++;
        return { valor: dato, contexto: { estabaVacia } };
    }

    // Eliminar la primera ocurrencia de "dato"
    eliminar(dato) {
        const estabaVacia = this.estaVacia();
        if (estabaVacia) {
            return { valor: dato, contexto: { estabaVacia } };
        }

        // Caso especial: el dato está en la cabeza
        if (this.cabeza.dato === dato) {
            this.cabeza = this.cabeza.siguiente;
            this.cantidad--;
            return { valor: dato, contexto: { estabaVacia, esCabeza: true, encontrado: true } };
        }

        // Buscamos el nodo anterior al que queremos eliminar
        let anterior = this.cabeza;
        while (anterior.siguiente !== null && anterior.siguiente.dato !== dato) {
            anterior = anterior.siguiente;
        }

        if (anterior.siguiente === null) {
            // No se encontró el valor
            return { valor: dato, contexto: { estabaVacia, esCabeza: false, encontrado: false } };
        }

        anterior.siguiente = anterior.siguiente.siguiente;
        this.cantidad--;
        return { valor: dato, contexto: { estabaVacia, esCabeza: false, encontrado: true } };
    }

    estaVacia() {
        return this.cabeza === null;
    }

    tamano() {
        return this.cantidad;
    }

    // Entrega todos los valores en orden, listos para graficar
    obtenerElementos() {
        const elementos = [];
        let actual = this.cabeza;
        while (actual !== null) {
            elementos.push(actual.dato);
            actual = actual.siguiente;
        }
        return elementos;
    }
}
